src/sfizz/Tuning.cpp

```diff
@@ -44,25 +46,59 @@ struct Tuning::Impl {
         mappingFromParameters(defaultRootKey, defaultTuningFrequency)
     };
 
+    absl::optional<fs::path> scalaFile_;
+    fs::file_time_type modificationTime_ {};
+
     static constexpr int numKeys = Tunings::Tuning::N;
     static constexpr int keyOffset = 256; // Surge tuning has key range ±256
     std::array<float, numKeys> keysFractional12TET_;
 };
 
-///
-constexpr int Tuning::Impl::defaultRootKey;
-constexpr float Tuning::Impl::defaultTuningFrequency;
-constexpr int Tuning::Impl::numKeys;
+void Tuning::Impl::reset()
+{
+    rootKey_ = defaultRootKey;
+    tuningFrequency_ = defaultTuningFrequency;
+    tuning_ = Tunings::Tuning(
+        Tunings::evenTemperament12NoteScale(),
+        mappingFromParameters(defaultRootKey, defaultTuningFrequency)
+    );
+    scalaFile_.reset();
+    modificationTime_ = fs::file_time_type::min();
+    updateKeysFractional12TET();
+}
 
 float Tuning::Impl::getKeyFractional12TET(int midiKey) const
 {
     return keysFractional12TET_[std::max(0, std::min(numKeys - 1, midiKey + keyOffset))];
 }
 
-void Tuning::Impl::updateScale(const Tunings::Scale& scale)
+void Tuning::Impl::updateScale(const Tunings::Scale& scale, absl::optional<fs::path> sourceFile)
 {
     tuning_ = Tunings::Tuning(scale, tuning_.keyboardMapping);
     updateKeysFractional12TET();
+
+    if (sourceFile) {
+        std::error_code ec;
+        modificationTime_ = fs::last_write_time(*sourceFile, ec);
+        scalaFile_ = sourceFile;
```

Comment:
It should assign `scaleFile_` regardless if the condition holds true or not.
It should stop file checking in the case that a file-based scala is replaced with another non-file-based.

Reply:
true